smac/epm/base_epm.py

```diff
@@ -12,64 +15,182 @@
+            X_feats = self.scaler.transform(X_feats)
+            X_feats = self.pca.transform(X_feats)
+            X = np.hstack((X[:, :self.n_params], X_feats ))
+        
+        return self._predict(X)
+    
+    def _predict(self, X:np.ndarray):
+        '''
+        Predict means and variances for given X.
 
-        instance_features : list
-            list of instance features
+        Parameters
+        ----------
+        X : np.ndarray of shape = [n_samples, n_features (config + instance
+        features)]
 
         Returns
         -------
-        predictions
+        means : np.ndarray of shape = [n_samples, 1]
+            Predictive mean
+        vars : np.ndarray  of shape = [n_samples, 1]
+            Predictive variance
         '''
         raise NotImplementedError()
     
-    def predict_marginalized_over_instances(self, X):
-        """Predict mean and variance marginalized over all instances.
+    def predict_marginalized_over_instances(self, X:np.ndarray):
+        '''Predict mean and variance marginalized over all instances.
 
         Returns the predictive mean and variance marginalised over all
         instances for a set of configurations.
 
         Parameters
         ----------
-        X : np.ndarray of shape = [n_features (config), ]
+        X : np.ndarray of shape = [n_samples, n_features (config + instance
+        features)]
 
         Returns
         -------
         means : np.ndarray of shape = [n_samples, 1]
             Predictive mean
         vars : np.ndarray  of shape = [n_samples, 1]
             Predictive variance
-        """
+        '''
+
+        if self.instance_features is None or \
+                len(self.instance_features) == 0:
+            mean, var = self.predict(X)
+            var[var < self.var_threshold] = self.var_threshold
+            var[np.isnan(var)] = self.var_threshold
+            return mean, var
+        else:
+            n_instance_features = self.instance_features.shape[1]
+            n_instances = len(self.instance_features)
+
+        if len(X.shape) != 2:
+            raise ValueError(
+                'Expected 2d array, got %dd array!' % len(X.shape))
+
+        mean = np.zeros(X.shape[0])
+        var = np.zeros(X.shape[0])
+        for i, x in enumerate(X):
+            X_ = np.hstack(
+                (np.tile(x, (n_instances, 1)), self.instance_features))
+            means, vars = self.predict(X_)
+            # use only mean of variance and not the variance of the mean here
+            # since we don't want to reason about the instance hardness distribution
+            var_x = np.mean(vars) # + np.var(means)
+            if var_x < self.var_threshold:
+                var_x = self.var_threshold
+
+            var[i] = var_x
+            mean[i] = np.mean(means)
+
+        var[var < self.var_threshold] = self.var_threshold
+        var[np.isnan(var)] = self.var_threshold
+
+        if len(mean.shape) == 1:
```

Comment:
If you set the shape correctly when initialising with numpy.zero, you don't need this here